includes/Story_Renderer.php

```diff
@@ -325,6 +335,75 @@ protected function replace_body_end_tag( $content ) {
 		return str_replace( '</body>', $output . '</body>', $content );
 	}
 
+	/**
+	 * Get the configuration object to use.
+	 *
+	 * @return Configuration Optimizer configuration to use.
+	 */
+	private function get_optimizer_configuration() {
+		$transformers = [
+			AmpSchemaOrgMetadata::class,
+			AmpRuntimeCss::class,
+			TransformedIdentifier::class,
+			ReorderHead::class,
+		];
+
+		$configuration = [ Configuration::KEY_TRANSFORMERS => $transformers ];
+		$config        = new Configuration( $configuration );
+		$config->registerConfigurationClass(
+			AmpSchemaOrgMetadata::class,
+			AmpSchemaOrgMetadataConfiguration::class
+		);
+
+		return $config;
+	}
+
+	/**
+	 * Optimizes the resulting markup using AMP Optimizer if available.
+	 *
+	 * @param string $content Story markup.
+	 *
+	 * @return string Filtered content.
+	 */
+	protected function optimize_markup( $content ) {
+		if ( ! class_exists( '\AmpProject\Optimizer\TransformationEngine' ) ) {
+			return $content;
+		}
+
+		$enable_optimizer = true;
+
+		/** This filter is documented in amp-wp/includes/class-amp-theme-support.php */
+		$enable_optimizer = apply_filters( 'amp_enable_optimizer', $enable_optimizer );
+
+		if ( ! $enable_optimizer ) {
+			return $content;
+		}
+
+		// Needed for \AmpProject\Optimizer\Transformer\AmpRuntimeCss.
+		$content = str_replace( '</head>', '<style amp-runtime></style></head>', $content );
+
+		$errors         = new ErrorCollection();
+		$optimizer      = new TransformationEngine( $this->get_optimizer_configuration() );
+		$optimized_html = $optimizer->optimizeHtml( $content, $errors );
+
+		if ( count( $errors ) === 0 ) {
+			return $optimized_html;
+		}
+
+		$error_messages = array_map(
+			static function( Error $error ) {
+				return ' - ' . $error->getCode() . ': ' . $error->getMessage();
+			},
+			iterator_to_array( $errors )
+		);
+
+		return str_replace(
+			'</html>',
+			"\n" . '<!---' . "\n" . 'AMP optimization could not be completed due to the following:' . "\n" . implode( "\n", $error_messages ) . "\n" . '-->' . "\n" . '</html>',
+			$content
+		);
```

Comment:
Wouldn't it be safer to append this to `$content` rather than try to inject it? In theory, there could be `<!--</html>-->` in the page.